kernel/thread.c

```diff
@@ -601,10 +600,9 @@ char *z_setup_new_thread(struct k_thread *new_thread,
 #endif
 #ifdef CONFIG_USERSPACE
 	/* New threads inherit any memory domain membership by the parent */
-	if (_current->mem_domain_info.mem_domain != NULL) {
-		k_mem_domain_add_thread(_current->mem_domain_info.mem_domain,
-					new_thread);
-	}
+	new_thread->mem_domain_info.mem_domain = NULL;
```

Comment:
This violates the 'a thread always has a memory domain'. Guess it is ok since this is during initialization and the domain is set in the next line - just wonder if there could be any corner cases/race conditions due to this temporary NULL. Change to default domain instead?

Reply:
later in this function we call k_mem_domain_add_thread() to perform the assignment. I can move this line immediately before it to make things clearer.

Reply:
when this happens the thread object is not in a completely initialized state and it will never be running. the NULL assignment here is to make sure k_mem_domain_add_thread() does the right thing (i.e. doesn't try to remove the thread from some other domain). This should be OK AFAIK. It wouldn't be correct to put `new_thread->mem_domain_info.mem_domain = &k_mem_domain_default` here.